extensions/amp-google-assistant-assistjs/0.1/assistjs-frame-service.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview The shared service used by all custom elements to talk to Assistant services. It loads an iframe that exports
+ * endpoints to handle requests from custom elements.
+ */
+
+import {Services} from '../../../src/services';
+import {addAttributesToElement} from '../../../src/dom';
+import {toggle} from '../../../src/style';
+
+export class AssistjsFrameService {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private @const {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = ampdoc;
+
+    /** @private {?AssistjsConfigService} */
+    this.configService_ = null;
+
+    /** Create Assistant iframe and append it to the main AMP document. */
+    this.createAssistantIframe_();
+  }
+
+  /** @private */
+  createAssistantIframe_() {
+    this.configService_ = Services.assistjsConfigServiceForDoc(this.ampDoc_);
+    const iframe = this.ampDoc_.getRootNode().createElement('iframe');
+    this.configService_.getWidgetIframeUrl('frame').then((iframeUrl) => {
+      addAttributesToElement(iframe, {
+        src: iframeUrl,
+        allow: 'microphone',
+      });
+      toggle(iframe, false);
+      document.body.appendChild(iframe);
```

Comment:
This **must** wait for the document to become visible, and should only be done in a `layoutCallback`.

Reply:
Maybe I miss something here but why does it need to wait for the entire document become visible? I thought we are just appending an element to the document, which could be done anytime even before the DOM tree is constructed entirely? If we want to put it in layoutCallback, which element should we put it in since it's a shared service across elements?

Reply:
Actually, I don't see any callers of `assistjsFrameServiceForDoc`. What is this service being used for?

Reply:
This PR is just creating the skeleton. The clients will be the three new components (`amp-google-assistant-voice-button`, `amp-google-assistant-voice-bar`, and `amp-google-assistant-inline-suggestion-bar`)

Reply:
Yes was trying to create a skeleton code first to build on, but also agree that might seem strange to add a service that no other component is calling. And a shared service with iframe doesn't look like(or related to) other base elements at all so that maybe it shouldn't be included in the very first PR of skeleton code. Let me move that into next PR with real usage of it.